### DISCUSSION TITLE
PhotoPrism container cannot connect to MariaDB service

### DISCUSSION THREAD

**Discussion by Kubinetes** [photoprism/photoprism#4347] [Q&A]
#### 1. What is not working as documented?

The PhotoPrism container is unable to establish a connection with the MariaDB service. As a result, the application cannot access the database, leading to a failure in functionality.

Getting these errors in the logs:

```
time="2024-06-20T23:16:13+02:00" level=info msg="config: waiting for the database to become available"
time="2024-06-20T23:20:13+02:00" level=error msg="dial tcp 10.0.9.16:3306: i/o timeout"
```

I checked if the MariaDB service is running correectly:

```
2024-06-19 23:54:00 0 [Note] Server socket created on IP: '0.0.0.0'.
2024-06-19 23:54:00 0 [Note] Server socket created on IP: '::'.
2024-06-19 23:54:00 0 [Note] mariadbd: Event Scheduler: Loaded 0 events
2024-06-19 23:54:00 0 [Note] mariadbd: ready for connections.
```

#### 2. How can we reproduce it?

Deploy `Docker-Compose.yml` in Portainer to Dockerswarm

#### 3. What behavior do you expect?

The PhotoPrism container should successfully connect to the MariaDB service, allowing the application to function normally with database access.

#### 4. What could be the cause of your problem?

At first I thought it was a networking issue, but when going in the container I was able to ping the service's IP:

```
root@240531:/photoprism$ ping 10.0.9.16
PING 10.0.9.16 (10.0.9.16): 56 data bytes
64 bytes from 10.0.9.16: icmp_seq=0 ttl=64 time=0.079 ms
64 bytes from 10.0.9.16: icmp_seq=1 ttl=64 time=0.082 ms
64 bytes from 10.0.9.16: icmp_seq=2 ttl=64 time=0.079 ms
64 bytes from 10.0.9.16: icmp_seq=3 ttl=64 time=0.068 ms
64 bytes from 10.0.9.16: icmp_seq=4 ttl=64 time=0.058 ms
64 bytes from 10.0.9.16: icmp_seq=5 ttl=64 time=0.067 ms
64 bytes from 10.0.9.16: icmp_seq=6 ttl=64 time=0.064 ms
64 bytes from 10.0.9.16: icmp_seq=7 ttl=64 time=0.071 ms
64 bytes from 10.0.9.16: icmp_seq=8 ttl=64 time=0.066 ms
64 bytes from 10.0.9.16: icmp_seq=9 ttl=64 time=0.074 ms
64 bytes from 10.0.9.16: icmp_seq=10 ttl=64 time=0.071 ms
64 bytes from 10.0.9.16: icmp_seq=11 ttl=64 time=0.075 ms
64 bytes from 10.0.9.16: icmp_seq=12 ttl=64 time=0.064 ms
64 bytes from 10.0.9.16: icmp_seq=13 ttl=64 time=0.066 ms
64 bytes from 10.0.9.16: icmp_seq=14 ttl=64 time=0.066 ms
64 bytes from 10.0.9.16: icmp_seq=15 ttl=64 time=0.066 ms
64 bytes from 10.0.9.16: icmp_seq=16 ttl=64 time=0.076 ms
64 bytes from 10.0.9.16: icmp_seq=17 ttl=64 time=0.065 ms
64 bytes from 10.0.9.16: icmp_seq=18 ttl=64 time=0.067 ms
64 bytes from 10.0.9.16: icmp_seq=19 ttl=64 time=0.053 ms
64 bytes from 10.0.9.16: icmp_seq=20 ttl=64 time=0.067 ms
64 bytes from 10.0.9.16: icmp_seq=21 ttl=64 time=0.067 ms
64 bytes from 10.0.9.16: icmp_seq=22 ttl=64 time=0.056 ms
^C64 bytes from 10.0.9.16: icmp_seq=23 ttl=64 time=0.050 ms
64 bytes from 10.0.9.16: icmp_seq=24 ttl=64 time=0.074 ms
^C--- 10.0.9.16 ping statistics ---
25 packets transmitted, 25 packets received, 0% packet loss
round-trip min/avg/max/stddev = 0.050/0.068/0.082/0.000 ms
```

#### 5. Can you provide us with example files for testing, error logs, or screenshots?

The Docker-Compose.yml I used: https://pastebin.com/DEa9iERS

Used Pastebin or else the post would become to long.

#### 6. Which software versions do you use?

(a) PhotoPrism Architecture & Build Number: **Latest Docker image**

(b) Database Type & Version: **MariaDB 11**

(c) Operating System Types & Versions: **Ubuntu 22.04 LTS underlaying host for Docker.**

**Comment by graciousgrey** [comment]
Did you use the docker-compose.yml example we do provide for portainer installations?

- https://dl.photoprism.app/docker/portainer/
- https://docs.photoprism.app/getting-started/portainer/

Our troubleshooting checklist for connection problems can be found here:
- https://docs.photoprism.app/getting-started/troubleshooting/mariadb/#cannot-connect

Please note that our examples are not made/optimized for docker swarm. 

In [this discussion](https://github.com/photoprism/photoprism/discussions/3945) a user also had a problem with docker swarm and connecting to maria db. He assumed that this might be related to an error in the ufw-setup.

**Reply by Kubinetes** [reply]
Yes I used the docker-compose.yml examples from the documentation.



I will check in on the other disscusion thanks for providing the link!

**Reply by lastzero** [reply]
What's definitely strange then is the private class A subnet (10.x.x.x), since Docker doesn't use it by default. I would thus assume there's some custom configuration in place...

**Reply by Kubinetes** [reply]
I think that's how the Docker overlay drivers work, since I have multiple stacks running also using the overlay driver and having no issues.
